optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc

```diff
@@ -58,55 +45,283 @@ that uses a functional approach:
 However, that scales poorly because it doesn't do an <<incrementalScoreCalculation,incremental calculation>>:
 When the planning variable of a single `Talk` changes, to recalculate the score,
 the normal Streams API has to execute the entire stream from scratch.
+The ConstraintStreams API allows you to write similar code in pure Java, while reaping the performance benefits of
+incremental score calculation.
+This is an example of the same code, using the Constraint Streams API:
 
-On the other hand, the ConstraintStreams API uses <<incrementalScoreCalculation,deltas>>:
+// TODO Rewrite the above paragraph PLANNER-1709 is fixed, mentioning constraint justification as another benefit.
 
-////
-TODO Replace the previous sentence when PLANNER-1709 is fixed.
+[source,java,options="nowrap"]
+----
+    private Constraint doNotAssignAnn(ConstraintFactory factory) {
+        return factory.from(Shift.class)
+                .filter(Shift::isEmployeeAnn)
+                .penalize("Don't assign Ann", HardSoftScore.ONE_SOFT);
+    }
+----
 
-So it doesn't do <<incrementalScoreCalculation,deltas>>.
-Nor does it support <<explainingTheScore,justifications>>.
+This constraint stream goes over all instances of class `Shift` amongst <<problemFacts,problem facts>> and
+<<planningEntity,planning entities>> in the <<planningProblemAndPlanningSolution,planning problem>>.
+It finds all which are assigned to employee `Ann` and for every such instance (also called "a match"), it adds a soft
+penalty of 1 to the overall <<calculateTheScore,score>>.
+The following figure illustrates this process on a problem with 4 different shifts:
+
+image::ConstraintStreams/constraintStreamIntroduction.png[align="center"]
+
+If any of the instances change during solving, the constraint stream automatically detects it and only recalculates
+the minimum necessary portion of the problem that is affected by the change.
+The following figure illustrates this <<incrementalScoreCalculation,incremental score calculation>>:
+
+image::ConstraintStreams/constraintStreamIncrementalCalculation.png[align="center"]
 
-On the other hand, the ConstraintStreams API uses <<incrementalScoreCalculation,deltas>>
-and supports <<explainingTheScore,justifications>>:
 ////
+TODO Include and describe the following image when PLANNER-1709 is fixed.
+image::ConstraintStreams/constraintStreamJustification.png[align="center"]
+////
+
+
+[[constraintStreamsFirst]]
+== Your First Constraint Stream
```

Comment:
I'd call this "Configuration" instead, with an id of "constraintStreamsConfiguration".

This isn't a guide or a blog post. This is a reference manual. That's a minor difference, but "your first" is more for guides and blog posts. Not sure. Wdyt @emmurphy1 ?